flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinatorTest.java

```diff
@@ -105,54 +102,147 @@ void testCloseAfterRequestIsReceived() throws Exception {
+                assertThat(serverSideProcess).isNotDone();
+
+                // wait for the connection to be established
+                socketServer.waitForConnectionToBeEstablished();
+                // in order to close the connection from the server's side
+                socketServer.closeAcceptingSocket();
+
+                assertEmptyResponseGeneratedFromClient(responseFuture, version);
+
+                assertThat(serverSideProcess).isNotDone();
 
-            // server closes here
-            failedResponseFuture =
+                dataFuture.complete(Collections.singletonList(Row.of(123, "abc")));
+
+                FlinkAssertions.assertThatFuture(serverSideProcess)
+                        .eventuallyFailsWith(ExecutionException.class)
+                        .withCauseInstanceOf(SocketException.class);
+            }
+        }
+    }
+
+    @Test
+    void testServerSideNotResponding() throws Exception {
+        try (final TestingSocketServer socketServer = new TestingSocketServer()) {
+            final String expectedVersion = "version";
+            final CompletableFuture<CoordinationResponse> responseFuture;
+            try (CollectSinkOperatorCoordinator coordinator =
+                    new CollectSinkOperatorCoordinator()) {
+                coordinator.start();
+                coordinator.handleEventFromOperator(
+                        0, 0, new CollectSinkAddressEvent(socketServer.getSocketAddress()));
+
+                responseFuture =
+                        coordinator.handleCoordinationRequest(
+                                createRequestForClientGeneratedResponse(expectedVersion));
+                assertThat(responseFuture).isNotDone();
+            }
+            assertEmptyResponseGeneratedFromClient(responseFuture, expectedVersion);
+        }
+    }
+
+    @Test
+    void testServerSideDisconnectWithReconnect() throws Exception {
+        try (CollectSinkOperatorCoordinator coordinator = new CollectSinkOperatorCoordinator()) {
+            coordinator.start();
+
+            final TestingSocketServer socketServer =
+                    TestingSocketServer.createSocketServerAndInitializeCoordinator(coordinator);
+
+            final String expectedVersion = "version";
+            final CompletableFuture<CoordinationResponse> responseFuture =
                     coordinator.handleCoordinationRequest(
-                            createRequestForClientGeneratedResponse(versionOfFailedRequest));
+                            createRequestForClientGeneratedResponse(expectedVersion));
+
+            final CompletableFuture<SocketConnection> connectFuture = socketServer.connectSocket();
+            FlinkAssertions.assertThatFuture(connectFuture)
+                    .as("The connection to the client should be established.")
+                    .eventuallySucceeds();
+
+            // simulation of the server side failing while the request is handled on the server side
```

Comment:
Maybe we should drop the client/server terminology because afaict we have 3 participants; the actual client (e.g., CLI), the coordinator and the sink operator (== TestingSocketServer), creating 2 client/server pairs.
It's just really confusing.

Reply:
As mentioend [above](https://github.com/apache/flink/pull/23296#discussion_r1308280794):
> I refactored the method and variable names to use coordinator and sinkFunction. Additionally, TestingSocketServer is renamed into TestingSinkFunction